pilot/pkg/xds/ads_test.go

```diff
@@ -43,6 +49,74 @@ const (
 	routeB = "https.443.https.my-gateway.testns"
 )
 
+type clientSecrets struct {
+	secretmodel.SecretItem
+}
+
+func (sc *clientSecrets) GenerateSecret(ctx context.Context, connectionID, resourceName, token string) (*secretmodel.SecretItem, error) {
+	return &sc.SecretItem, nil
+}
+
+// ShouldWaitForIngressGatewaySecret indicates whether a valid ingress gateway secret is expected.
+func (sc *clientSecrets) ShouldWaitForIngressGatewaySecret(connectionID, resourceName, token string, fileMountedCertsOnly bool) bool {
+	return false
+}
+
+// TODO: must fix SDS, it uses existence to detect it's an ACK !!
+func (sc *clientSecrets) SecretExist(connectionID, resourceName, token, version string) bool {
+	return false
+}
+
+// DeleteSecret deletes a secret by its key from cache.
+func (sc *clientSecrets) DeleteSecret(connectionID, resourceName string) {
+}
+
+// TestAgent will start istiod with TLS enabled, use the istio-agent to connect, and then
+// use the ADSC to connect to the agent proxy.
+func TestAgent(t *testing.T) {
+	// Start Istiod
+	bs, tearDown := initLocalPilotTestEnv(t)
+	defer tearDown()
+
+	// TODO: when authz is implemented, verify labels are checked.
+	cert, key, err := bs.CA.GenKeyCert([]string{"spiffe://cluster.local/fake.test"}, 1*time.Hour)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	creds := &clientSecrets{
+		secretmodel.SecretItem{
+			PrivateKey:       key,
+			CertificateChain: cert,
+			RootCert:         bs.CA.GetCAKeyCertBundle().GetRootCertPem(),
+		},
+	}
+
+
+	t.Run("adscTLSDirect", func(t *testing.T) {
+		testAdscTLS(t, creds)
+	})
+
+}
+
+// testAdscTLS tests that ADSC helper can connect using TLS to Istiod
+func testAdscTLS(t *testing.T, creds cache.SecretManager) {
+	// connect to the local XDS proxy - it's using a transient port.
+	ldsr, err := adsc.Dial(util.MockPilotSGrpcAddr, "",
+		&adsc.Config{
+			IP:        "10.11.10.1",
+			Namespace: "test",
+			Secrets:   creds,
+			Watch: []string{
+				v3.ClusterType,
+				collections.IstioNetworkingV1Alpha3Serviceentries.Resource().GroupVersionKind().String()},
+		})
+	if err != nil {
+		t.Fatal("Failed to connect", err)
```

Comment:
is just dial sufficient? and it would fail if tls is not set up correctly?

Reply:
Yes, it fails - I spent far more time than I should have to get it to pass ( localhost and SANs not matching, etc). 

We can add more tests here - the larger PR has few more checks - but I was thinking we should move the other tests to use TLS, to be closer to reality.